src/main/java/org/dependencytrack/tasks/maintenance/ProjectMaintenanceTask.java

```diff
@@ -0,0 +1,126 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) OWASP Foundation. All Rights Reserved.
+ */
+package org.dependencytrack.tasks.maintenance;
+
+import alpine.common.logging.Logger;
+import alpine.event.framework.Event;
+import alpine.event.framework.Subscriber;
+import org.dependencytrack.event.maintenance.ProjectMaintenanceEvent;
+import org.dependencytrack.persistence.jdbi.ConfigPropertyDao;
+import org.dependencytrack.persistence.jdbi.ProjectDao;
+import org.jdbi.v3.core.Handle;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static net.javacrumbs.shedlock.core.LockAssert.assertLocked;
+import static org.dependencytrack.model.ConfigPropertyConstants.MAINTENANCE_PROJECTS_RETENTION_DAYS;
+import static org.dependencytrack.model.ConfigPropertyConstants.MAINTENANCE_PROJECTS_RETENTION_ENABLE;
+import static org.dependencytrack.model.ConfigPropertyConstants.MAINTENANCE_PROJECTS_RETENTION_TYPE;
+import static org.dependencytrack.model.ConfigPropertyConstants.MAINTENANCE_PROJECTS_RETENTION_VERSIONS;
+import static org.dependencytrack.persistence.jdbi.JdbiFactory.inJdbiTransaction;
+import static org.dependencytrack.persistence.jdbi.JdbiFactory.openJdbiHandle;
+import static org.dependencytrack.persistence.jdbi.JdbiFactory.withJdbiHandle;
+import static org.dependencytrack.util.LockProvider.executeWithLock;
+import static org.dependencytrack.util.TaskUtil.getLockConfigForTask;
+
+public class ProjectMaintenanceTask implements Subscriber {
+
+    private static final Logger LOGGER = Logger.getLogger(ProjectMaintenanceTask.class);
+
+    @Override
+    public void inform(final Event event) {
+        if (!(event instanceof ProjectMaintenanceEvent)) {
+            return;
+        }
+
+        final long startTimeNs = System.nanoTime();
+        try {
+            LOGGER.info("Starting project maintenance");
+            final Statistics statistics = executeWithLock(
+                    getLockConfigForTask(ProjectMaintenanceTask.class),
+                    () -> informLocked());
+            if (statistics == null) {
+                LOGGER.info("Task is locked by another instance; Skipping");
+                return;
+            }
+
+            final var taskDuration = Duration.ofNanos(System.nanoTime() - startTimeNs);
+            LOGGER.info("Completed in %s: %s".formatted(taskDuration, statistics));
+        } catch (Throwable e) {
+            final var taskDuration = Duration.ofNanos(System.nanoTime() - startTimeNs);
+            LOGGER.error("Failed to complete after %s".formatted(taskDuration), e);
+        }
+    }
+
+    private record Statistics(int deletedInactiveProjects) {
+    }
+
+    private Statistics informLocked() {
+        assertLocked();
+        AtomicInteger numDeletedTotal = new AtomicInteger(0);
+
+        final Handle jdbiHandle = openJdbiHandle();
```

Comment:
This must be wrapped in a try-with-resources, or you should use `withJdbiHandle` instead (preferable). If anything fails between this, and `jdbiHandle.close()`, the backing DB connection would remain open.

Reply:
At the moment, the handle is also never closed when `MAINTENANCE_PROJECTS_RETENTION_ENABLE` is `false`, for example.